src/encoder/mod.rs

```diff
@@ -0,0 +1,425 @@
+        for (tag, (field_type, count, offset)) in self.ifd.iter() {
+            self.writer.write_u16(*tag)?;
+            self.writer.write_u16(*field_type)?;
+            self.writer.write_u32(*count)?;
+            self.writer.write_bytes(offset)?;
+        }
+
+        Ok(offset)
+    }
+
+    /// Write some data to the tiff file, the offset of the data is returned.
+    ///
+    /// This could be used to write tiff strips.
+    pub fn write_data<T: TiffValue>(&mut self, value: T) -> TiffResult<u64> {
+        let offset = self.writer.offset();
+        value.write(&mut self.writer)?;
+        Ok(offset)
+    }
+
+    fn finish_internal(&mut self) -> TiffResult<()> {
+        let ifd_pointer = self.write_directory()?;
+        let curr_pos = self.writer.offset();
+
+        self.writer.goto_offset(self.ifd_pointer_pos)?;
+        self.writer.write_u32(ifd_pointer as u32)?;
+        self.writer.goto_offset(curr_pos)?;
+        self.writer.write_u32(0)?;
+
+        self.dropped = true;
+
+        Ok(())
+    }
+
+    /// Write out the ifd directory.
+    pub fn finish(mut self) -> TiffResult<()> {
+        self.finish_internal()
+    }
+}
+
+impl<'a, W: Write + Seek> Drop for DirectoryEncoder<'a, W> {
+    fn drop(&mut self) {
+        if !self.dropped {
+            let _ = self.finish_internal();
+        }
+    }
+}
+
+
+/// Type to encode images strip by strip.
+///
+/// You should call `finish` on this when you are finished with it.
+/// Encoding can silently fail while this is dropping.
+///
+/// # Examples
+/// ```
+/// # extern crate tempfile;
+/// # let mut file = tempfile::tempfile().unwrap();
+/// # let mut image_data = Vec::new();
+/// # for x in 0..100 {
+/// #     for y in 0..100u8 {
+/// #         let val = x + y;
+/// #         image_data.push(val);
+/// #         image_data.push(val);
+/// #         image_data.push(val);
+/// #     }
+/// # }
+/// use tiff::encoder::*;
+///
+/// let mut tiff = TiffEncoder::new(&mut file).unwrap();
+/// let mut image = tiff.new_image::<colortype::RGB8>(100, 100).unwrap();
+///
+/// let mut idx = 0;
+/// while image.next_strip_sample_count() > 0 {
+///     let sample_count = image.next_strip_sample_count() as usize;
+///     image.write_strip(&image_data[idx..idx+sample_count]).unwrap();
+///     idx += sample_count;
+/// }
+/// image.finish().unwrap();
+/// ```
+pub struct ImageEncoder<'a, W: Write + Seek, C: ColorType> {
+    encoder: DirectoryEncoder<'a, W>,
+    strip_idx: u64,
+    strip_count: u64,
+    row_samples: u64,
+    height: u32,
+    rows_per_strip: u64,
+    strip_offsets: Vec<u32>,
+    strip_byte_count: Vec<u32>,
+    dropped: bool,
+    _phantom: std::marker::PhantomData<C>,
+}
+
+impl<'a, W: Write + Seek, T: ColorType> ImageEncoder<'a, W, T> {
+    fn new(mut encoder: DirectoryEncoder<'a, W>, width: u32, height: u32) -> TiffResult<ImageEncoder<'a, W, T>> {
+        let row_samples = width as u64 * <T>::bits_per_sample().len() as u64;
```

Comment:
There are a few remaining clippy warning regarding casts in this method and the next one.

Reply:
Will fix